source/AI.cpp

```diff
@@ -1589,12 +1589,16 @@ void AI::MoveIndependent(Ship &ship, Command &command) const
 		return;
 	}
 
-	// A ship has restricted movement options if it is 'staying' or is hostile to its parent.
+	// A ship has restricted movement options if it is 'staying' or is hostile to its parent,
+	// or has the lingering personality.
+	const System *origin = ship.GetSystem();
 	const bool shouldStay = ship.GetPersonality().IsStaying()
-			|| (ship.GetParent() && ship.GetParent()->GetGovernment()->IsEnemy(gov));
+			|| (ship.GetParent() && ship.GetParent()->GetGovernment()->IsEnemy(gov))
+			|| (origin && !ship.IsFleeing() && ship.GetPersonality().IsLingering() &&
+			Random::Int(max<int>(300, origin->MinimumFleetPeriod())));
+
```

Comment:
The existing parts at least fit on a single line each.
```suggestion
			|| (ship.GetParent() && ship.GetParent()->GetGovernment()->IsEnemy(gov));
	if(ship.GetPersonality().IsLingering() && !ship.IsFleeing())
		shouldStay |= (origin && Random::Int(max<int>(300, origin->MinimumFleetPeriod())));

```

Reply:
The `origin` check should be before `IsLingering` and `!IsFleeing` since `origin` just requires a comparison to 0, while the other two require setting up a call stack.

Reply:
Likewise with `shouldStay`. It should be checked before calling five functions.

Reply:
```suggestion
			|| (ship.GetParent() && ship.GetParent()->GetGovernment()->IsEnemy(gov));
	if(origin && ship.GetPersonality().IsLingering() && !ship.IsFleeing())
		shouldStay = shouldStay || Random::Int(max<int>(300, origin->MinimumFleetPeriod()));

```

Reply:
You're still doing three function calls before checking `shouldStay` (GetPersonality, IsLingering, and IsFleeing). The `shouldStay` belongs before those.

Reply:
```suggestion
			|| (ship.GetParent() && ship.GetParent()->GetGovernment()->IsEnemy(gov));
	if(!shouldStay && origin && ship.GetPersonality().IsLingering() && !ship.IsFleeing())
		shouldStay |= Random::Int(max<int>(300, origin->MinimumFleetPeriod()));

```
Though, it's not necessarily true that this is faster. This introduces additional branching and none of those functions are actually expensive, retrieving the personality object or the fleeing bools aren't difficult, and `IsLingering()` is just a bitwise AND.

Reply:
I just wrote a few paragraphs and code examples to prove you're wrong, but I'd rather stop arguing about this point, so I deleted them. You may have the last word.

Besides, I want to replace that massive mess with an equivalent else-if chain inside a method. It'll be a whole lot easier to understand.